Add step argument to SummaryWriter.(set_)as_default.

foxik opened the pull request:
As a followup to #26406, here is a simple proposed solution which adds `step` argument to `SummaryWriter.as_default` and `SummaryWriter.set_as_default`. The default value is `None` which means the argument is ignored, so the change is backward compatible.

The current code is extremely simple, yet useful. Given lack of response in #26406, I submitted this PR, but I am happy to update it (or scrap it) if a different consensus is reached. For reasons to this change, see #26406.

BTW, the `get_step` and `set_step` methods could be used inside the implementation instead of manual manipulation with `_summary_state`.

foxik committed:
Add step argument to SummaryWriter.(set_)as_default.

superbobry reviewed (dismissed):
I'm not sure why both `set_as_default` and `as_default` exist, but the PR itself looks good.

gbaned commented:
@foxik it is struck with internal error , can you please fix this? Thanks!

```
Traceback (most recent call last):
  File "<embedded stdlib>/unittest/case.py", line 59, in testPartExecutor
    yield
  File "<embedded stdlib>/unittest/case.py", line 605, in run
    testMethod()
  File "/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 410, in testAPIBackwardsCompatibility
    omit_golden_symbols_map=omit_golden_symbols_map)
  File "/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 387, in _checkBackwardsCompatibility
    api_version=api_version)
  File "/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 307, in _AssertProtoDictEquals
    self.fail('%d differences found between API and golden.' % diff_count)
  File "/third_party/py/absl/testing/absltest.py", line 1767, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
  File "<embedded stdlib>/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: 1 differences found between API and golden.

Traceback (most recent call last):
  File "<embedded stdlib>/unittest/case.py", line 59, in testPartExecutor
    yield
  File "<embedded stdlib>/unittest/case.py", line 605, in run
    testMethod()
  File "/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 443, in testAPIBackwardsCompatibilityV2
    omit_golden_symbols_map=omit_golden_symbols_map)
  File "/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 387, in _checkBackwardsCompatibility
    api_version=api_version)
  File "/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 307, in _AssertProtoDictEquals
    self.fail('%d differences found between API and golden.' % diff_count)
  File "/third_party/py/absl/testing/absltest.py", line 1767, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
  File "<embedded stdlib>/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: 1 differences found between API and golden.
```

foxik commented:
@gbaned I updated the API files and the test now passes. I modified the original commit, which invalidated the review -- should I in the future not modify commits that have been already reviewed and create new ones instead?

karmel commented on tensorflow/python/ops/summary_ops_v2.py:
```diff
@@ -197,14 +197,32 @@ class SummaryWriter(object):
+  def as_default(self, step=None):
+    """Returns a context manager that enables summary writing.
+
+    For the `step` argument, see `tf.summary.experimental.set_step()`.
```
Here and elsewhere: can we instead repeat the relevant sentences in these docstrings as well?

Also, can you add an example of how to use this argument?

foxik replied:
Done, I included all relevant information in the doc string, and added an example, which also shows that `step` is reverted when the context manager exits.

jsimsa replied:
@nfelt could you advise @foxik on how to add tests? Not having test coverage for public API is undesirable.

foxik replied:
The tests have been added.

karmel commented on tensorflow/python/ops/summary_ops_v2.py:
```diff
@@ -197,14 +197,32 @@ class SummaryWriter(object):
+    Args:
+      step: An `int64`-castable default step value, or `None`.
+        When not `None`, the current step is captured, replaced by a given one,
+        and the original one is restored when the context manager exists.
```
nit: when the context manager **exits**

foxik replied:
Thanks, fixed.

jsimsa replied:
@foxik thank you for the changes, please also add tests for this functionality

foxik replied:
@jsimsa I will try, but I am not sure about the required scope, as `tf.summary.experimental.{get,set}_step` nor `SummaryWriter`s do not have any tests.

Is it enough to check that after using `set_as_default` and `as_default`, the `get_step` returns what it should? (Plus restoration of the step after `as_default` exists?)

foxik replied:
Added tests that check the real generated event logs.

karmel reviewed (changes requested):
(for tensorflow/api-owners)

* Will tf.summary.experimental.set_step() be deprecated as part of this addition?
* This functionality should be tested.

foxik committed:
Include more information in the docstrings.

* Instead of referencing tf.summary.experimental.set_step, spell out the
  relevant information directly.
* Fix a typo
* Add example to the `as_default` method.

gbaned commented:
@jsimsa  Any update on this PR? Please. Thanks!

jsimsa commented:
@gbaned please work with @nfelt and @foxik on adding tests for the new functionality.

foxik commented:
@jsimsa @gbaned I was waiting for @nfelt guidance, but I will instead implement the tests by myself.

I just found https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/summary_ops_test.py, which performs SummaryWriter tests with all the functionality needed (i.e., loging a summary using a SummaryWriter and then reading back the `step` number). Should send something in a day or two.

foxik committed:
Tests for step parameter of SummaryWriter.(set_)as_default.

Four new tests are added:
- we test both constants and variables as step values
- we test as_default and set_as_default methods

For as_default, we also check that the original value is
restored if and only if a non-None step was passed.

foxik commented:
@jsimsa @gbaned I added tests that check both the `as_default` and `set_as_default` methods, checking both constants and variables as `step` values. Furthermore, for the `as_default` I check that the step value is restored if and only if a non-None `step` was passed.

foxik committed:
Make sure all lines are at most 80 characters long.

foxik commented:
I am sorry, but two lines in the tests were 81 characters long :-( I commited a fix.

foxik commented:
@gbaned Just a gentle reminder, the pull request seems ready (the requested changes on my part are implemented). Cheers!

gbaned commented:
> @gbaned Just a gentle reminder, the pull request seems ready (the requested changes on my part are implemented). Cheers!

@foxik  Sorry for delay. It is on internal processing.

foxik commented:
For curiosity, any chance this can get in before the TF 2.3 branch cut? Or maybe it is held back to allow more time to discuss public API changes?

nfelt commented:
Sorry I have been out of the loop on this PR, I haven't had bandwidth to work on the summary API in a while.  Thanks for your contribution in suggesting this change.

I'm not sure what the current thoughts of the TF API approvers are, but I'm a little reluctant to introduce the `step` argument as a permanent non-experimental part of the `as_default()`/`set_as_default()` API since we hadn't fully resolved some questions about how to handle step tracking in the long run, and I'm not sure that part of the API is the best place for it to live long-term [*].  I would be open to introducing it as something like `experimental_step` if that gives us flexibility to deprecate it before TF 3.0 if we do find an alternate approach, but I'm not sure if TF typically does experimental parameters like that?

[*]: Just to elaborate slightly, it's possible that we want to integrate step tracking a little tighter into the summary writers so that e.g. each writer maintains an internal step count, in which case it makes more sense to have something like a `writer.advance_step()` API rather than allowing `as_default()` to temporarily reassign the step.

foxik commented:
@nfelt Thanks for your comment! Having per-writer summary step would be great.

I understand that you want to keep open possibility for API changes/redesign. However, I do not think having a `step` argument of `as_default()/set_as_default()` would prevent that:
- `step` argument is already part of the individual `tf.summary` calls
- even if `step` is per-writer, it will surely be changeable

So the current pull request does not seem to introduce additional invariants or concepts to the API.

But maybe it would make sense to change the comments to something like
> The `as_default()/set_as_default()` is guaranteed to change the `step` argument only for the calls regarding the summary writer in question. Currently, it changes the global `step`, but it could change in the future.
